invoice-api/src/service/freshbooks.ts

```diff
@@ -0,0 +1,503 @@
+        this.emailInvoice(newInvoiceID,emailInfoJson);
+
+    };
+
+    private stripAdminEmails(emails:Array<any>, adminEmails:Array<any>): Array<string> {
+        let output = [];
+        // any email that is not an oicr admin is a valid customer email
+        for(var idx in emails){
+            if(adminEmails.indexOf(emails[idx].toLowerCase()) <0){
+                output.push(emails[idx]);
+            }
+        }
+        return output;
+
+    }
+
+    // assumes date is : null or in 'YYYY-MM-DD' format
+    public async getInvoicesSummaryData(date:string, user:any, admin:boolean) {
+        await this.authenticate();
+        let minDate = null;
+        let maxDate = null;
+        minDate = date;
+        maxDate = new Date();
+        let pageCount = 1; let pageIdx = 1;
+        let allInvoicesForThisPeriod = [];
+        let allCustomerIDs = null;
+        if(!admin){
+            console.log('Sending request to Freshbooks: For getting Customers');
+            let customersInfo = await this.getAllCustomers(user.email);
+            allCustomerIDs = this.getCustomerIDs(customersInfo);
+            if(allCustomerIDs.length == 0) throw Error("error: No account found for this user account: "+ user.email)
+        }
+
+        // currently we have decided to pull all the invoices and ship it to front end in one go
+        while(pageCount  >= pageIdx){
+            let invoiceResults =
+                await this.getInvoicesListPaged(pageIdx,minDate, maxDate.toISOString().slice(0,10), allCustomerIDs);
+            allInvoicesForThisPeriod = allInvoicesForThisPeriod.concat(invoiceResults.invoices);
+            //update page count as the first result tells total count of pages
+            if(pageCount == 1) pageCount = invoiceResults.pages;
+            pageIdx++;
+        }// while loop ends here
+        let flattenedInovicesJson = [];
+        _.each(allInvoicesForThisPeriod, (item) => {
+            flattenedInovicesJson.push({
+                'current_organization' : item.current_organization,
+                'date' : item.create_date,
+                'invoice_number' : item.invoice_number,
+                'payment_status' : item.payment_status,
+                'invoice_status' : item.v3_status,
+                'cpu_cost' : this.getLineItemValue("CPU",item.lines),
+                'image_cost' : this.getLineItemValue("Image",item.lines),
+                'volume_cost' : this.getLineItemValue("Volume",item.lines),
+                'discount' : item.discount_value,
+                'total' : item.amount.amount
+            });
+        });
+        return flattenedInovicesJson;
+    };
+
+    private async getAllCustomers(userEmail:string): Promise<any>{
+        userEmail = (userEmail == null || userEmail == "") ?  "":userEmail;
+        let pageCount = 1; let pageIdx = 1;
+        let allCustomersData = [];
+        while(pageCount  >= pageIdx){
```

Comment:
at first glance the while loop takes a second to understand. Another issue is that it has to make each request sequentially, waiting for the next one to come back before moving forward. 

A better way would be to just fire off all the requests you'll need in parallel, and maintain an array of promises which you can just `await` on with `Promise.all()`

Something like this (my syntax might be wrong):

```javascript
    private async getAllCustomersParallel(userEmail:string): Promise<any>{
        userEmail = (userEmail == null || userEmail == "") ?  "" : userEmail;
        let promises = [];
        let pagedCustomerResults = await this.getAllCustomersPaged(1, userEmail);
        
        for (let i=1; i<pagedCustomerResults.pages; i++) {
            promises.push(this.getAllCustomersPaged(i, userEmail));
        }
        
        let responses = await Promise.all(promises);
        return responses.reduce((a,b) => a.concat(b));
    }
```